### PR TITLE
Add Stratification class, tests, and docstrings

### PR DESCRIPTION
**Pull request by mattkappel**
### Description
- *Category*: feature
- *JIRA issue*: [MIC-3132](https://jira.ihme.washington.edu/browse/MIC-3132)

#### Changes
- Adds Stratification class and fixes small bug in the non-vectorized flow
- Adds docstrings to document the Stratification class
- Adds 10 pytest tests, for expected outputs and expected exception cases

### Testing
Ran new pytests successfully.

### CODE REVIEW

**Review by albrja — APPROVED** [review]

File: tests/framework/results/test_stratification.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
+import pandas as pd
+import pytest
+
+from vivarium.framework.results.stratification import Stratification
+
+##########################
+# Mock data and fixtures #
+##########################
+NAME = "hogwarts_house"
```
Comment (stevebachmeier) [inline tests/framework/results/test_stratification.py:10]: I've never seen the movies :( 

(though I have read a few of the first books!)

**Review by stevebachmeier — APPROVED** [review]

File: src/vivarium/framework/results/stratification.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        if self.mapper is None:
+            if len(self.sources) != 1:
+                raise ValueError(
+                    f"No mapper provided for stratification {self.name} with"
+                    f"{len(self.sources)} stratification sources."
+                )
+            self.mapper = self._default_mapper
+            self.is_vectorized = True
+
+    def __call__(self, population: pd.DataFrame) -> pd.DataFrame:
```
Comment (rmudambi) [inline src/vivarium/framework/results/stratification.py:49]: Should variable name be `population` or would somethign like `source_data` be better?

File: src/vivarium/framework/results/stratification.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        if self.is_vectorized:
+            raw_mapped_column = self.mapper(population[self.sources])
+        else:
+            raw_mapped_column = population[self.sources].apply(self.mapper, axis=1)
+        mapped_column = raw_mapped_column.astype(
+            CategoricalDtype(categories=self.categories, ordered=True)
+        )
+        if mapped_column.isna().any():
+            invalid_categories = set(raw_mapped_column.unique()) - set(self.categories)
+            raise ValueError(f"Invalid values '{invalid_categories}' found in {self.name}.")
```
Comment (rmudambi) [inline src/vivarium/framework/results/stratification.py:59]: Need to test this case

File: src/vivarium/framework/results/stratification.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    name: str
+    sources: List[str]
+    categories: List[str]
+    mapper: Callable[[Union[pd.Series, pd.DataFrame]], Union[str, pd.Series]] = None
+    is_vectorized: bool = False
+
+    def __post_init__(self):
+        if self.mapper is None:
+            if len(self.sources) != 1:
+                raise ValueError(
```
Comment (rmudambi) [inline src/vivarium/framework/results/stratification.py:42]: Need to test this failure case

File: src/vivarium/framework/results/stratification.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                    f"No mapper provided for stratification {self.name} with"
+                    f"{len(self.sources)} stratification sources."
+                )
+            self.mapper = self._default_mapper
+            self.is_vectorized = True
+
+    def __call__(self, population: pd.DataFrame) -> pd.DataFrame:
+        if self.is_vectorized:
+            raw_mapped_column = self.mapper(population[self.sources])
+        else:
```
Comment (rmudambi) [inline src/vivarium/framework/results/stratification.py:52]: Need some edge cases:
What if is_vectorized is inconsistent with the mapper?
What if categories is empty?
What if sources is empty?
What if sources is not in population.columns?

Spend some time thinking about other possible edge cases/failure cases

File: tests/framework/results/test_stratification.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+SOURCES = ["first_name", "last_name"]
+CATEGORIES = ["hufflepuff", "ravenclaw", "slytherin", "gryffindor"]
+STUDENT_TABLE = pd.DataFrame(
+    np.array([["harry", "potter"], ["severus", "snape"], ["luna", "lovegood"]]),
+    columns=SOURCES,
+)
+
+
+def sorting_hat_vector(state_table: pd.DataFrame) -> pd.Series:
```
Comment (rmudambi) [inline tests/framework/results/test_stratification.py:19]: Need to use the argument in this mapper

File: src/vivarium/framework/results/stratification.py (hunk 2 above)
Comment (mattkappel) [reply src/vivarium/framework/results/stratification.py:49]: We definitely use `population` in a lot of places in vivarium and refer to discrete elements of that population as "simulants". I think we should keep `population` for consistency. After having done a refactor to `source_data`, you've then got a problem of describing what the `source_data` is, and generally, that's a population with simulants (humans, maybe, but could be boids or something else).

File: src/vivarium/framework/results/stratification.py (hunk 3 above)
Comment (mattkappel) [reply src/vivarium/framework/results/stratification.py:59]: Agree will add!

File: src/vivarium/framework/results/stratification.py (hunk 4 above)
Comment (mattkappel) [reply src/vivarium/framework/results/stratification.py:42]: Agree will add!

File: src/vivarium/framework/results/stratification.py (hunk 5 above)
Comment (mattkappel) [reply src/vivarium/framework/results/stratification.py:52]: Agree will add several erroring cases, with possibly adding code to have the errors be handled properly!

Comment (mattkappel) [reply src/vivarium/framework/results/stratification.py:52]: > What if is_vectorized is inconsistent with the mapper?

This one is turning out to be a tricky case. It will almost certainly throw an exception, but in a way specific to that `mapper`'s implementation. I'll check for a generic `Exception` here.
